File: src/test/java/com/avast/kluzo/javaapi/KluzoTest.java
package com.avast.kluzo.javaapi;

import com.avast.continuity.ContinuityContextThreadNamer;
import com.avast.continuity.ThreadNamer;
import com.avast.continuity.javaapi.Continuity;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Optional;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import static org.junit.Assert.assertEquals;

public class KluzoTest {

    private Logger logger = LoggerFactory.getLogger(this.getClass());

    @Test
    public void testTraceId() throws InterruptedException {
        ThreadNamer threadNamer = ContinuityContextThreadNamer.prefix("traceId");
        ExecutorService executor = Executors.newCachedThreadPool();
        ExecutorService wrappedExecutor = Continuity.wrapExecutorService(executor, threadNamer);

        CountDownLatch latch = new CountDownLatch(3);

        TraceId traceId1 = TraceId.generate();
        Kluzo.withTraceId(traceId1, () -> {
            wrappedExecutor.execute(() -> {
                assertEquals(Kluzo.getTraceId(), Optional.of(traceId1));
                logger.info("first");
            });
            latch.countDown();
            return null;
        });

        TraceId traceId2 = TraceId.generate();
        Kluzo.withTraceId(traceId2, () -> {
            executor.execute(() -> {
                assertEquals(Kluzo.getTraceId(), Optional.empty());
                logger.info("second");
            });
            latch.countDown();
            return null;
        });

        TraceId traceId3 = TraceId.generate();
        Kluzo.withTraceId(traceId3, () -> {
            wrappedExecutor.execute(() -> {
                assertEquals(Kluzo.getTraceId(), Optional.of(traceId3));
                logger.info("third");
            });
            latch.countDown();
            return null;
        });

        latch.await();
        Thread.sleep(500);
    }

}

File: kluzo-scala/src/test/scala/com/avast/kluzo/KluzoTest.scala
package com.avast.kluzo

import java.util.concurrent.{Executor, Executors}

import com.avast.continuity.{Continuity, ContinuityContextThreadNamer}
import com.avast.kluzo.Kluzo._
import org.scalatest.FunSuite
import org.slf4j.LoggerFactory

import scala.concurrent.ExecutionContext

class KluzoTest extends FunSuite {

  /** See the log output to check that traceId is present. */
  private val logger = LoggerFactory.getLogger(this.getClass)

  test("traceId") {
    testTraceId(Executors.newSingleThreadExecutor)
    testTraceId(ExecutionContext.global)
  }

  private def testTraceId(pool: Executor): Unit = {
    val wrappedPool = Continuity.wrapExecutor(pool)(ContinuityContextThreadNamer.prefix(ContinuityKey))

    val traceId1 = TraceId.generate
    Kluzo.withTraceId(Some(traceId1)) {
      wrappedPool.execute(new Runnable {
        override def run(): Unit = {
          assert(getTraceId === Some(traceId1))
          logger.info("first")
        }
      })
    }

    val traceId2 = TraceId.generate
    Kluzo.withTraceId(Some(traceId2)) {
      pool.execute(new Runnable {
        override def run(): Unit = {
          assert(getTraceId === None)
          logger.info("second")
        }
      })
    }

    val traceId3 = TraceId.generate
    Kluzo.withTraceId(Some(traceId3)) {
      wrappedPool.execute(new Runnable {
        override def run(): Unit = {
          assert(getTraceId === Some(traceId3))
          logger.info("third")
        }
      })
    }
  }

  ignore("performance") {
    // implicit val tn = ContinuityContextThreadNamer.prefix(ContinuityKey)
    val tid = TraceId("test")
    val count = 100000
    var totalTime = 0L
    var workTime = 0L
    for (i <- 0 until count) {
      val start = System.nanoTime
      Kluzo.withTraceId(Some(tid)) {
        val start1 = System.nanoTime
        var sum = 0L
        for (j <- i until 100000) {
          if (j % 2 == 0) {
            sum += j
          } else {
            sum *= j
          }
        }
        workTime += System.nanoTime - start1
      }
      totalTime += System.nanoTime - start
    }

    println(s"workTime: $workTime, average: ${ workTime.toDouble / count }")
    println(s"totalTime: $totalTime, average: ${ totalTime.toDouble / count }")
    println(s"overhead: ${ totalTime - workTime }, average: ${ (totalTime - workTime).toDouble / count }")
  }

}
